borsh/src/schema_helpers.rs

```diff
@@ -36,3 +36,208 @@ pub fn schema_container_of<T: BorshSchema>() -> BorshSchemaContainer {
+    }
+
+    match defs.get(declaration).ok_or(declaration) {
+        Ok(Definition::Array { length, elements }) => {
+            // Aggregate `count` and `length` to a single number.  If this
+            // overflows, check if array’s element is zero-sized.
+            let count = usize::try_from(*length)
+                .ok()
+                .and_then(|len| len.checked_mul(count));
+            match count {
+                Some(0) => Some(0),
+                Some(count) => max_serialized_size_impl(count, elements, defs),
+                None if is_zero_size(elements, defs) => Some(0),
+                None => None,
+            }
+        }
+        Ok(Definition::Sequence { elements }) => {
+            // Assume that sequence has MAX_LEN elements since that’s the most
+            // it can have.
+            let sz = max_serialized_size_impl(MAX_LEN, elements, defs)?;
+            // Add four to account for encoded length.
+            count.checked_mul(sz.checked_add(4)?)
+        }
+
+        Ok(Definition::Enum { variants }) => {
+            // Size of an enum is the largest variant plus one for tag.
+            let mut max = 0;
+            for (_, variant) in variants {
+                let sz = max_serialized_size_impl(1, variant, defs)?;
+                max = max.max(sz);
+            }
+            max.checked_add(1)
+        }
+
+        // Tuples and structs sum sizes of all the members.
+        Ok(Definition::Tuple { elements }) => tuple(count, elements, defs),
+        Ok(Definition::Struct { fields }) => match fields {
+            Fields::NamedFields(fields) => {
+                tuple(count, fields.into_iter().map(|(_, field)| field), defs)
+            }
+            Fields::UnnamedFields(fields) => tuple(count, fields, defs),
+            Fields::Empty => Some(0),
+        },
+
+        // Handle primitive types.  They have well-known sizes.
+        Err("nil") => Some(0),
+        Err("bool" | "i8" | "u8") => Some(count),
+        Err("i16" | "u16") => count.checked_mul(2),
+        Err("i32" | "u32" | "f32") => count.checked_mul(4),
+        Err("i64" | "u64" | "f64") => count.checked_mul(8),
+        Err("i128" | "u128") => count.checked_mul(16),
+
+        // Assume string with maximum length.  This is equivalent to `Vec<u8>`.
+        Err("string") => count.checked_mul(MAX_LEN.checked_add(4)?),
+
+        Err(_) => None,
+    }
+}
+
+#[test]
+fn test_max_serialized_size() {
+    #[cfg(not(feature = "std"))]
+    use alloc::string::String;
+
+    mod test_structs {
+        use crate as borsh;
+        #[cfg(not(feature = "std"))]
+        use alloc::string::ToString as _;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestEmpty;
+
+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestNamed {
+            _foo: usize,
+            _bar: [u8; 15],
+        }
+
+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestUnnamed(usize, [u8; 15]);
```

Comment:
adding a test for recursive struct onto current [commit](https://github.com/mina86/borsh-rs/tree/8ae9f6c580790548509b64b4dda80ef2a51438d7) results in a stack overflow:

```diff
adding a `TestRecursive` struct test
diff --git a/borsh/src/schema_helpers.rs b/borsh/src/schema_helpers.rs
index 017e5c9f..4e4a38b2 100644
--- a/borsh/src/schema_helpers.rs
+++ b/borsh/src/schema_helpers.rs
@@ -193,20 +193,26 @@ fn test_max_serialized_size() {

         #[derive(::borsh_derive::BorshSchema)]
         pub struct TestEmpty;

         #[derive(::borsh_derive::BorshSchema)]
         pub struct TestNamed {
             _foo: usize,
             _bar: [u8; 15],
         }

+        #[derive(::borsh_derive::BorshSchema)]
+        pub struct TestRecursive {
+            a: u64,
+            b: Box<Option<TestRecursive>>,
+        }
+
         #[derive(::borsh_derive::BorshSchema)]
         pub struct TestUnnamed(usize, [u8; 15]);
     }

     const MAX_LEN: usize = 4294967295;

     assert_eq!(Some(2), max_serialized_size::<u16>());
     assert_eq!(Some(8), max_serialized_size::<usize>());

     assert_eq!(Some(1), max_serialized_size::<Option<()>>());
@@ -233,11 +239,12 @@ fn test_max_serialized_size() {
         Some(0),
         max_serialized_size::<[[[(); MAX_LEN]; MAX_LEN]; MAX_LEN]>()
     );

     use test_structs::*;

     assert_eq!(Some(0), max_serialized_size::<TestEmpty>());
     assert_eq!(Some(23), max_serialized_size::<TestNamed>());
     assert_eq!(Some(23), max_serialized_size::<TestUnnamed>());
     assert_eq!(None, max_serialized_size::<BorshSchemaContainer>());
+    assert_eq!(None, max_serialized_size::<TestRecursive>());
 }

```
=> 

```bash
# test error `cargo test --features schema` in `./borsh`
thread 'schema_helpers::test_max_serialized_size' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass `--lib`

```

Depending on the way to proceed decided upon, a way to detect recursion may be added to `max_serialized_size` implementation.